Sources/SemanticIndex/UpdateIndexStoreTaskDescription.swift

```diff
@@ -475,29 +634,89 @@ package struct UpdateIndexStoreTaskDescription: IndexTaskDescription {
+    struct TargetAndLanguage: Hashable, Comparable {
+      let target: BuildTargetIdentifier
+      let language: Language
+
+      static func < (lhs: TargetAndLanguage, rhs: TargetAndLanguage) -> Bool {
+        if lhs.target.uri.stringValue < rhs.target.uri.stringValue {
+          return true
+        } else if lhs.target.uri.stringValue > rhs.target.uri.stringValue {
+          return false
+        }
+        if lhs.language.rawValue < rhs.language.rawValue {
+          return true
+        } else if lhs.language.rawValue > rhs.language.rawValue {
+          return false
+        }
+        return false
+      }
+    }
+
+    var partitions: [(target: BuildTargetIdentifier, language: Language, files: [FileIndexInfo])] = []
+    var fileIndexInfosToBatch: [TargetAndLanguage: [FileIndexInfo]] = [:]
+    for fileIndexInfo in fileIndexInfos {
+      guard fileIndexInfo.language == .swift,
+        await buildServerManager.toolchain(for: fileIndexInfo.target, language: fileIndexInfo.language)?
+          .canIndexMultipleSwiftFilesInSingleInvocation ?? false
+      else {
+        // Only Swift supports indexing multiple files in a single compiler invocation, so don't batch files of other
+        // languages.
+        partitions.append((fileIndexInfo.target, fileIndexInfo.language, [fileIndexInfo]))
+        continue
+      }
+      // Even for Swift files, we can only index files in a single compiler invocation if they have the same build
+      // settings (modulo some normalization in `updateIndexStore(forFiles:)`). We can't know that all the files do
+      // indeed have the same compiler arguments but loading build settings during scheduling from the build server
+      // is not desirable because it slows down a bottleneck.
+      // Since Swift files within the same target should build a single module, it is reasonable to assume that they all
+      // share the same build settings.
+      let languageAndTarget = TargetAndLanguage(target: fileIndexInfo.target, language: fileIndexInfo.language)
+      fileIndexInfosToBatch[languageAndTarget, default: []].append(fileIndexInfo)
+    }
+    let batchedPartitions =
+      fileIndexInfosToBatch
+      .sorted { $0.key < $1.key }  // Ensure we get a deterministic partition order
+      .flatMap { targetAndLanguage, files in
+        // The batch size of 5 was chosen without too many significant performance measurements because most projects
+        // currently indexed by SourceKit-LSP are limited by preparation time instead of indexing time and it's thus
+        // hard to quanify the performance characteristics of different batch sizes. 5 seems like a good trade-off to
+        // share work between files within the same target without overloading a single job with too many files and
+        // thus loosing parallelism.
+        files.partition(intoBatchesOfSize: 5).map { (targetAndLanguage.target, targetAndLanguage.language, $0) }
```

Comment:
Could make configurable? What's the driver choose for its batch sizes?

Reply:
Good idea to check with the driver. It schedules as many concurrent batches as the machine has CPU cores but limits the batch size to 25 (https://github.com/swiftlang/swift-driver/blob/main/Sources/SwiftDriver/Jobs/Planning.swift#L917-L1022). I think copying that behavior is as reasonable an ideas as I have. But I’d prefer to do that in a follow-up PR to keep this one focused on the actual implementation of batch indexing instead of introducing complications due to configuration options.